### ISSUE TITLE
Remove Generic "Expression Can Result in Runtime Error" Diagnostic

### ISSUE THREAD

**Issue by JeanMertz** [vectordotdev/vector#8925]
ref #8922

#### Pain

The "expression can result in runtime error" diagnostic is an error message that
occurs when a root level expression is considered fallible:

```coffee
to_string(.foo)
```

```
error[E100]: unhandled error
  ┌─ :1:1
  │
1 │ to_string(.foo)
  │ ^^^^^^^^^^^^^^^
  │ │
  │ expression can result in runtime error
  │ handle the error case to ensure runtime success
```

The above example is fairly straightforward, but when the number of nested
expressions gets deeper, it becomes more difficult to understand what is
happening:

```coffee
if (parsed  = parse_grok(.message, "%{GREEDYDATA:parsed}"); parsed != null) {
  merge(., parsed)
}
```

The same error is raised when you try to divide by zero:

```coffee
1 / 0
```

Or when you group multiple expression types:

```coffee
5 + to_int(.foo)
```

The issue we need to solve, is that we can't know by looking at a single
expression whether or not the root level expression will be fallible.

Take this example:

```coffee
to_int(.foo)
```

On its own, `to_int(.foo)` is fallible, because we don't know the eventual value
of `.foo` at runtime, and because `to_int` can be fallible for certain input
types, we don't know if `to_int` can fail at compile-time.

However, when compounding the function call expression with an error-coalescing
expression, we can make the root level expression infallible:

```coffee
to_int(.foo) ?? 0
```

Now, if `to_int` were to fail, it would default to `0`, and thus the root-level
expression becomes infallible.

Because of this, we have a catch-all "expression can result in runtime error"
error, which is used when we can't determine if an individual expression itself
needs to be infallible for the program to compile, but we _can_ determine if
each individual root expression (the combination of all expressions on a given
source code line) is infallible or not.

The problem is, this error is ambiguous, and gives no clear direction on how to
solve the problem.

Take this example:

```coffee
"foo" + .bar + baz[1]
```

```
error[E100]: unhandled error
  ┌─ :1:1
  │
1 │ "foo" + .bar + baz[1]
  │ ^^^^^^^^^^^^^^^^^^^^^
  │ │
  │ expression can result in runtime error
  │ handle the error case to ensure runtime success
```

It is not immediately clear which part of this program is fallible. It's either
assigning `.bar.` to `"foo"`, or `baz[1]` to `"foo" + .bar`.

This is still a fairly simple example, but you can make it as complex as you
want, making it more and more difficult to understand what needs fixing.

#### User Experience

Whenever fallibility of an expression prevents a program from compiling, the
compiler must always point to the exact expression that causes the problem,
instead of highlighting the entire chain of expressions.

#### Implementation

Given the following nested expressions:

```
A ( B ( C ( D ) ) )
```

Where:

- `A` is the root expression
- `C` is fallible
- `D` is infallible

The compiler must return a diagnostic similar to this:

```
error[E100]: unhandled error
  ┌─ :1:1
  │
1 │ A ( B ( C ( D ) ) )
  │         ^^^^^^^
  │         │
  │         C expression is fallible
  │         handle the error case to ensure runtime success
```

Here's what the compiler will do:

- Compile the nested expressions, from `D` to `A`
- `D` is infallible, and thus valid
- `C` is fallible
  - But it might become infallible by a parent expression
  - Set `C` as the source of fallibility for this expression chain
- `B` is fallible
- `A` is fallible
  - This is the root expression, and thus the fallibility of `C` remains
    unhandled
- Show the error diagnostic specific to the unhandled fallibility of `C`

The compiler has a `compile_root_exprs` function:

```rust
fn compile_root_exprs(
    &mut self,
    nodes: impl IntoIterator<Item = Node<ast::RootExpr>>,
) -> Vec<Expr> {
    // ...
}
```

This function is called first to compile a VRL program. For each root
expression, it does something like the following:

```rust
for expr in root_expressions {
    let expr = self.compile_expr(expr);
    if expr.is_fallible() {
        self.errors.push(ExpressionError);
    }
}
```

This will need to be changed. Instead, we'll add an extra piece of state to the
compiler, with the error to show _if_ a root expression remains infallible,
something like:

```rust
for expr in root_expressions {
    self.fallible_expression_error = None;
    let expr = self.compile_expr(expr);
    if let Some(error) = self.fallible_expression_error {
        self.errors.push(error);
    }
}
```

The `fallible_expression_error` state will be set by the first expression in the
chain (going from last to first) that is infallible, and will be set back to
`None` by any subsequent expression that is infallible (indicating that the
fallibility of the child expression is nullified).

If any expression further up the chain is marked as fallible again, the value
becomes `Some(T)` again, until we've reached the root expression, which is
handled by the above code snippet.

This allows us to be specific about which part of a VRL program is fallible, and
needs to be rectified.

**Comment by JeanMertz** [comment]
There's a new generic diagnostic we want to get rid of as well: https://github.com/vectordotdev/vector/pull/11172/files#r799419317

**Comment by JeanMertz** [comment]
see also: https://github.com/vectordotdev/vector/pull/6594#discussion_r587427961

**Comment by JeanMertz** [comment]
Resolved in #12880.
